### ISSUE TITLE
nodeAffinity retryStrategy children node lookup is done on name while 'Children(parentId)' expects ID

### ISSUE THREAD

**Issue by heyleke** [argoproj/argo-workflows#5830]
## Summary 

Using retryStrategy nodeAntiAffinity to work as example describes 
```yaml
    retryStrategy:
      limit: 10
      retryPolicy: "Always"
      affinity:
        nodeAntiAffinity: {}
```

## Diagnostics

Seen on private K8S cluster

tested on v3.0.1 (but checked master branch as well)

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: workflow-
  name: wf-test
spec:
  entrypoint: workflow-example
  templates:
  - name: workflow-example
    steps:
    - - name: run
        template: sleepabit
        continueOn:
          failed: true
  - name: sleepabit
    retryStrategy:
      limit: 1
      retryPolicy: "Always"
      affinity:
        nodeAntiAffinity: {}
    container:
      image: alpine:3.7
      command: [sh, -c]
      args: [sleep 15 && false"]
```

logs of controller with some extra prints shoul that the lookup of the children fails, and so hostNames is empty:
```
time="2021-05-05T12:51:38.776Z" level=info msg="Get leases 200"
time="2021-05-05T12:51:38.782Z" level=info msg="Processing workflow" namespace=default workflow=wf-test
time="2021-05-05T12:51:38.783Z" level=info msg="Pod failed: Error (exit code 1)" displayName="run(0)" namespace=default pod=wf-test-73877493 templateName=sleepabit workflow=wf-test
[...]
time="2021-05-05T12:51:38.789Z" level=info msg="jheylen: retry" namespace=default workflow=wf-test
time="2021-05-05T12:51:38.789Z" level=info msg="1 child nodes of wf-test[0].run failed. Trying again..." namespace=default workflow=wf-test
time="2021-05-05T12:51:38.789Z" level=info msg="jheylen: create an new child" namespace=default workflow=wf-test
time="2021-05-05T12:51:38.789Z" level=info msg="jheylen: retry on different host" namespace=default workflow=wf-test
time="2021-05-05T12:51:38.790Z" level=info msg="seen retryStrategy: {1 Always nil &RetryAffinity{NodeAntiAffinity:&RetryNodeAntiAffinity{},}}"
time="2021-05-05T12:51:38.790Z" level=info msg="jheylen: seen nodes: map[]"
time="2021-05-05T12:51:38.790Z" level=info msg="jheylen: seen failNodes: map[]"
time="2021-05-05T12:51:38.790Z" level=info msg="seen hostNames: []"
time="2021-05-05T12:51:38.790Z" level=info msg="Pod node wf-test-2020228392 initialized Pending" namespace=default workflow=wf-test
time="2021-05-05T12:51:38.795Z" level=info msg="Create events 201"
time="2021-05-05T12:51:38.800Z" level=info msg="Create pods 201"
time="2021-05-05T12:51:38.801Z" level=info msg="Created pod: wf-test[0].run(1) (wf-test-2020228392)" namespace=default workflow=wf-test
time="2021-05-05T12:51:38.801Z" level=info msg="Workflow step group node wf-test-4187779359 not yet completed" namespace=default workflow=wf-test
time="2021-05-05T12:51:38.860Z" level=info msg="Update workflows 200"
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by goock** [comment]
> /review @goock

Could be related to this? 
https://github.com/argoproj/argo-workflows/issues/5054

**Comment by alexec** [comment]
@goock did you do write the code for this originally?

**Comment by heyleke** [comment]
> > /review @goock
> 
> Could be related to this?
> #5054

I have a second pr waiting where I fetch the hostNodeName sooner (at any podstate update), but wanted to be sure that the way I'm handling this issue and pr correctly

**Comment by goock** [comment]
> @goock did you do write the code for this originally?

I introduced the feature. I'll take a look.

**Comment by goock** [comment]
`retryNodeName` seems to be the node ID, however the code from #5831 ensures we always get node ID, originally the `retryNodeName` comes from https://github.com/argoproj/argo-workflows/blob/d66954f5b9b09e030408483502b03aa29727039a/workflow/controller/operator.go#L1501
https://github.com/argoproj/argo-workflows/blob/d66954f5b9b09e030408483502b03aa29727039a/workflow/controller/operator.go#L1661

@heyleke do you have the other PR opened?

**Comment by heyleke** [comment]
I don't fully get what you want to say with your comment, can you rephrase?

on the other PR for the sooner fetching of hostNodeName, no I didn't create the PR yet, will do it

**Comment by heyleke** [comment]
see #5858 for draft PR on the hostNodeName potential fix
